From 9dc95caf54a9830c097e73e8da512d194495f0bb Mon Sep 17 00:00:00 2001
From: "siyong.chen" <siyong.chen@sophgo.com>
Date: Wed, 13 May 2026 09:17:19 +0800
Subject: [PATCH 68/84] [fix](venc) fix sbm chn destroy no need check skip

[description ]:
[root cause  ]: NA
[jira id     ]: NA
[chip project]: cv184x
[side effects]: NA
[Affected]:

Signed-off-by: siyong.chen <siyong.chen@sophgo.com>
Change-Id: Id8acb5bc476f5604ada515b070e119a71200d927
---
 interdrv/vc_drv/driver/enc_api/src/enc_impl.c |  4 +--
 interdrv/vc_drv/driver/include/venc.h         |  1 +
 interdrv/vc_drv/driver/venc.c                 | 35 +++++++------------
 3 files changed, 15 insertions(+), 25 deletions(-)

diff --git a/interdrv/vc_drv/driver/enc_api/src/enc_impl.c b/interdrv/vc_drv/driver/enc_api/src/enc_impl.c
index ea112237..96e6200f 100644
--- a/interdrv/vc_drv/driver/enc_api/src/enc_impl.c
+++ b/interdrv/vc_drv/driver/enc_api/src/enc_impl.c
@@ -1542,9 +1542,9 @@ int vc_drop_sbm_frame_and_pskip(struct vc_enc_inst_s *enc_impl, struct hal_enc_t
 	reg &= ~(1 << 14);
 	vc_lite_mmu_write_reg(REG_SBM, 0x00, reg);
 	if (enc_impl->coding != CODEC_JPEG)
-		vc_sbm_irq_disable(SBM_IRQ_JPEGE_FRAME_DOEN_BIT);
-	else
 		vc_sbm_irq_disable(SBM_IRQ_FRAME_DOEN_BIT);
+	else
+		vc_sbm_irq_disable(SBM_IRQ_JPEGE_FRAME_DOEN_BIT);
 	vc_sbm_irq_enable(SBM_IRQ_PUSH_BIT);
 	return 0;
 }
diff --git a/interdrv/vc_drv/driver/include/venc.h b/interdrv/vc_drv/driver/include/venc.h
index 3acb7e94..54004b6d 100644
--- a/interdrv/vc_drv/driver/include/venc.h
+++ b/interdrv/vc_drv/driver/include/venc.h
@@ -145,6 +145,7 @@ typedef struct _venc_chn_context {
 	osal_wait wait_queue;
 	osal_wait wait_sbm_done;
 	unsigned int is_encoding;
+	unsigned int vpss_sbm_rdy;
 } venc_chn_context;
 
 typedef struct _venc_sbm_context {
diff --git a/interdrv/vc_drv/driver/venc.c b/interdrv/vc_drv/driver/venc.c
index a133a85a..bba8dd8d 100644
--- a/interdrv/vc_drv/driver/venc.c
+++ b/interdrv/vc_drv/driver/venc.c
@@ -2267,6 +2267,10 @@ static int _drv_init_chn_ctx(venc_chn ve_chn, const venc_chn_attr_s *pst_attr)
 	venc_chn_context *p_chn_handle;
 	venc_enc_ctx *p_enc_ctx;
 
+	if (g_venc_chns_ctx->chn_handle[ve_chn] != NULL) {
+		DRV_VENC_ERR("Channel %d has been created !\n", ve_chn);
+		return DRV_ERR_VENC_EXIST;
+	}
 	g_venc_chns_ctx->chn_handle[ve_chn] = osal_vzalloc(sizeof(venc_chn_context));
 	if (get_chn_handle(ve_chn) == NULL) {
 		DRV_VENC_ERR("Allocate chn_handle memory failed !\n");
@@ -3126,7 +3130,7 @@ int wait_smb_condition(const void *param)
 {
 	int condition;
 	venc_chn_context *p_chn_handle = (venc_chn_context *)param;
-	condition = !p_chn_handle->is_encoding;
+	condition = !p_chn_handle->is_encoding && !p_chn_handle->vpss_sbm_rdy;
 	return (condition > 0);
 }
 
@@ -3950,27 +3954,7 @@ int drv_venc_destroy_chn(venc_chn ve_chn)
 
 	if (p_chn_handle->sbm_state != VENC_SBM_STATE_IDLE
 		&& p_sbm_handle->p_sbm_enc_thread) {
-		venc_chn_status_s st_stat;
-		venc_stream_s st_stream = {0};
-		int is_not_need_skip = 0;
-
 		osal_mutex_lock(&p_sbm_handle->sbm_mutex);
-		is_not_need_skip = venc_sbm_get_skip_frm_status(&p_chn_handle->st_sbm_setting);
-		if (!is_not_need_skip) {
-			ret = drv_venc_cb_skip_frame(ve_chn, 1000);
-			if (ret == 0) {
-				ret = drv_venc_query_status(ve_chn, &st_stat);
-				st_stream.pst_pack = (venc_pack_s *)osal_vmalloc(sizeof(venc_pack_s) *
-					st_stat.u32_cur_packs);
-				if (st_stream.pst_pack) {
-					ret = drv_venc_get_stream(ve_chn, &st_stream, 1000);
-					if (ret == 0)
-						drv_venc_release_stream(ve_chn, &st_stream);
-					if (st_stream.pst_pack)
-						osal_vfree(st_stream.pst_pack);
-				}
-			}
-		}
 		if (p_chn_handle->p_chn_attr->st_venc_attr.en_type == PT_JPEG)
 			p_sbm_handle->jpeg_en = 0;
 		p_chn_handle->sbm_state = VENC_SBM_STATE_CHN_CLOSED;
@@ -6271,6 +6255,7 @@ static int venc_sbm_send_frame_thread(void *data)
 		ret = drv_venc_send_frame(p_chn_handle->ve_chn, &p_chn_handle->st_video_frame_info, -1);
 		p_chn_handle->is_encoding = 0;
 		p_sbm_handle->need_reset = 1;
+		p_chn_handle->vpss_sbm_rdy = 0;
 		venc_cb_vpss_sbm_frm_done((void*)&p_sbm_handle->curr_sbm_chn);
 		{
 			venc_chn_context *next_chn_handle = NULL;
@@ -6348,7 +6333,7 @@ int drv_venc_set_sbm_info(int venc_chn, struct sbm_cfg *sbm_info, int s32_milli_
 		DRV_VENC_INFO("venc chn no create\n");
 		return -1;
 	}
-	p_sbm_handle->curr_sbm_chn = venc_chn;
+
 	if (p_chn_handle->p_chn_attr->st_venc_attr.en_type == PT_H265 ||
 		p_chn_handle->p_chn_attr->st_venc_attr.en_type == PT_H264) {
 		st_sbm_setting.codec = 0x1;
@@ -6388,7 +6373,10 @@ int drv_venc_set_sbm_info(int venc_chn, struct sbm_cfg *sbm_info, int s32_milli_
 		p_sbm_handle->sbm_clk_en = 1;
 	}
 
-	vc_set_sbm_reg(&st_sbm_setting, 0);
+	if (!p_sbm_handle->p_sbm_enc_thread) {
+		p_sbm_handle->curr_sbm_chn = venc_chn;
+		vc_set_sbm_reg(&st_sbm_setting, 0);
+	}
 	osal_memcpy(&p_chn_handle->st_sbm_setting, &st_sbm_setting,
 		sizeof(struct venc_sbm_setting_s));
 
@@ -6632,6 +6620,7 @@ int drv_venc_cb_switch_chn(int venc_chn, u8 snr_num, u64 frm_num)
 		osal_spin_unlock(&p_sbm_handle->sbm_spin);
 		return -1;
 	}
+	p_chn_handle->vpss_sbm_rdy = 1;
 	osal_spin_unlock(&p_sbm_handle->sbm_spin);
 	if (p_sbm_handle->need_reset) {
 		reset_sbm(1);
-- 
2.25.1

